lib/rdoc/parser/prism_ruby.rb

```diff
@@ -0,0 +1,975 @@
+      @scanner.add_alias_method(node.old_name.value.to_s, node.new_name.value.to_s, node.location.start_line)
+    end
+
+    def visit_module_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+      module_name = constant_path_string(node.constant_path)
+      mod = @scanner.add_module_or_class(module_name, node.location.start_line, node.location.end_line) if module_name
+      if mod
+        @scanner.with_container(mod) do
+          super
+          @scanner.process_comments_until(node.location.end_line)
+        end
+      else
+        @scanner.skip_comments_until(node.location.end_line)
+      end
+    end
+
+    def visit_class_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+      superclass_name = constant_path_string(node.superclass) if node.superclass
+      class_name = constant_path_string(node.constant_path)
+      klass = @scanner.add_module_or_class(class_name, node.location.start_line, node.location.end_line, is_class: true, superclass_name: superclass_name) if class_name
+      if klass
+        @scanner.with_container(klass) do
+          super
+          @scanner.process_comments_until(node.location.end_line)
+        end
+      else
+        @scanner.skip_comments_until(node.location.end_line)
+      end
+    end
+
+    def visit_singleton_class_node(node)
+      @scanner.process_comments_until(node.location.start_line - 1)
+
+      expression = node.expression
+      expression = expression.body.body.first if expression.is_a?(Prism::ParenthesesNode) && expression.body&.body&.size == 1
+
+      case expression
+      when Prism::ConstantWriteNode
+        # Accept `class << (NameErrorCheckers = Object.new)` as a module which is not actually a module
+        mod = @scanner.container.add_module(RDoc::NormalModule, expression.name.to_s)
+      when Prism::ConstantPathNode, Prism::ConstantReadNode
+        expression_name = constant_path_string(expression)
+        # If a constant_path does not exist, RDoc creates a module
+        mod = @scanner.find_or_create_module_path(expression_name, :module) if expression_name
+      when Prism::SelfNode
+        mod = @scanner.container if @scanner.container != @top_level
+      end
+      if mod
+        @scanner.with_container(mod, singleton: true) do
+          super
+          @scanner.process_comments_until(node.location.end_line)
+        end
+      else
+        @scanner.skip_comments_until(node.location.end_line)
+      end
+    end
+
+    def visit_def_node(node)
+      start_line = node.location.start_line
+      end_line = node.location.end_line
+      @scanner.process_comments_until(start_line - 1)
+
+      case node.receiver
+      when Prism::NilNode, Prism::TrueNode, Prism::FalseNode
+        visibility = :public
+        singleton = false
+        receiver_name = { Prism::NilNode => 'NilClass', Prism::TrueNode => 'TrueClass', Prism::FalseNode => 'FalseClass' }[node.receiver.class]
```

Comment:
Could this trigger new hash allocation every time it's called?

Reply:
Changed to `case` `when` :+1: